### ISSUE TITLE
What exactly makes certain fixes unsafe should be explained

### ISSUE THREAD

**Issue by sh-at-cs** [astral-sh/ruff#8868]
It would be cool if Ruff gave an explanation for *why* certain fixes are considered unsafe, either in the docs or (even better) in the CLI output.

E.g. in the case of the fix for [UP007](https://docs.astral.sh/ruff/rules/non-pep604-annotation/), one can guess that it's because some libraries read annotations at runtime and might behave differently for `Union[X | Y]` and `X | Y` (just a guess, maybe that's wrong), but AFAICT, nothing like that is mentioned anywhere in the docs nor in the CLI output.

Maybe there could be a mandatory section explaining the relevant fixes' safety status in the docs page of each rule that has any.

**Comment by dhruvmanila** [comment]
Yeah, I agree. This is related to https://github.com/astral-sh/ruff/issues/8482. Perhaps for the CLI, we could implement it using https://github.com/astral-sh/ruff/issues/4361, which requires the reason to be mentioned in the code. Subsequently, the documentation could extract the message from there, as is currently done for the diagnostic message.
